### ISSUE TITLE
Cannot open include file: 'stdalign.h': No such file or directory

### ISSUE THREAD

**Issue by boris-999** [Cyan4973/xxHash#666]
MSVC 2019,  in a C++17/C17 project.

xxhash.h(1135,1): fatal error C1083: Cannot open include file: 'stdalign.h': No such file or directory

**Comment by Cyan4973** [comment]
`C11` mandates the existence of `<stdalign.h>`,
and `MSVC2019` is `C11` compliant,

hence chances are that `xxhash` was not compiled in `C` mode.

Let's now assume that the code is compiled in `C++` mode,
that means the preprocessor is nonetheless selecting the `C` path,
which is guarded by the following test : 
`if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L) /* >= C11 */`

I'm not familiar enough to know if it's normal / expected for a compiler in C++ mode to also claim compliance with `C11` while simultaneously not providing one of its required library headers.
The first part of the sentence is unclear, but the second part seems like a bug to me.

Either way, this might be something which could be circumvented by testing `C++` mode first.

But what would be really useful is to have a CI test that reproduce this issue,
so that we can observe it being fixed,
and ensure it remains so in the future.

**Comment by t-mat** [comment]
To address this issue, I've tested the following code as `main.c` or `main.cpp`.

```C
#include <stdio.h>

#if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L)
#  include <stdalign.h>
#endif

int main() {
#if defined(__STDC_VERSION__)
    printf("__STDC_VERSION__=%d\n", __STDC_VERSION__);
#else
    printf("__STDC_VERSION__ is not defined\n");
#endif
}
```

With the following MSVCs

- Microsoft Visual Studio Community 2019, Version 16.11.11, Platform Toolset = v142
- Microsoft Visual Studio Community 2022, Version 17.1.3, Platform Toolset = v143

For details of MSVC compiler options for C standard, see ["Predefined macros"](https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros) and [`/std`](https://docs.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version).


## Result (`__STDC_VERSION__`)

|version	|source		|Legacy MSVC	|/std:c11	|/std:c17	|
| --		| --		| --		| --		| --		|
|MSVC 2019	|main.c		|not defined	|201112		|201710		|
|		|main.cpp	|not defined	|not defined	|not defined	|
|MSVC 2022	|main.c		|not defined	|201112		|201710		|
|		|main.cpp	|not defined	|not defined	|not defined	|

notes:

(1) For all MSVCs, C/C++ and C standards, `<stdalign.h>` is included properly.  There's no compile error for all configuration.
(2) For C++ code, MSVC doesn't define any `__STDC_VERSION__`.
(3) For C code, MSVC follows `/std` standards.
(4) For any C++ standards (C++14/17/20/latest), results are same.

@boris-999, since I didn't find any issue about `__STDC_VERSION__` and `<stdalign.h>`.  Could you please provide further details of your issue?

**Comment by boris-999** [comment]
it was a while ago, can't fully remember what i did.    something @Cyan4973 said must have helped, or i turned off some feature so i could continue..    here's a screenie from my project made today.. still C++17/C17 proj in msvc2019..   sorry about the contrast off the #ifdef/#ifndef   items,  but you can see teh 1st branch at line 1134  in not currently taken (as that's for C)  2nd branch is taken, that's for C++.   when i click on the 1135, #include <stdalign> and select Go To Doc (F12),   the error dialog box  in the attached image still appears
![r43t4t](https://user-images.githubusercontent.com/92807520/161436843-0c2215d5-506d-4997-a2f9-747c82faf2f7.png)
.

**Comment by t-mat** [comment]
@boris-999, I'm really sorry.  I don't understand your situation.  Obviously this is caused by my lack of ability, knowledge and experience.

But if you still have some patience, could you provide the following information to us?

(1) Which version of MSVC are you using?
You can retrieve this information via `MSVC 2019 Menu > Help > About Microsoft Visual Studio`.  Upper left corner of the dialog indicates the version.

(2) Which version of Platform toolset are you using?
You can see this information at `MSVC 2019 Menu > Project > Property > Configuration Properties > General > (Right Pane) > Platform Toolset`.

(3) Which version of Windows Kits are you using?  You can check this information by the following procedure.
- Open some file in your project which includes ucrt libarary header.  e.g. `<stdlib.h>`, `<stdio.h>`, `<math.h>`, etc.
- Right click the `#include` line such as `#include <stdio.h>`.
- Select "Go To Document <stdio.h> (F12)".
- Right click the tab of `stdio.h` and select "Copy Full Path"
- Typically, it's something like `C:\Program Files (x86)\Windows Kits\10\Include\XX.X.XXXXX.X\ucrt\stdio.h` and `XX.X.XXXXX.X` indicates the version of Windows Kit which you are using.

**Comment by boris-999** [comment]
msvc2019 - 16.11.11 (now.  might have been something less 4months ago when i originally posted this bug).
platform tookset: Visual Studio 2019 (v142).
(only other option in there is LLVM clang-cl)

had look thru my codebase, i can't find anywhere i'm #defining out the issue;  so i must have resolved it.  

found a <stdlib.h>  in xxhash src file..  open containing folder..   
cmd prompt:

C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\ucrt>dir std*.h

 Directory of C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\ucrt
25/11/2020  06:39 PM             1,413 stddef.h
25/11/2020  06:39 PM            81,489 stdio.h
25/11/2020  06:39 PM            43,968 stdlib.h
               3 File(s)        126,870 bytes


the Notes comments here  looks vaguely intereseting:
https://en.cppreference.com/w/cpp/header/cstdalign

some googling implies MSVC doesn't have some C11 files..   one of the links:
https://github.com/zenny-chen/simple-stdatomic-for-VS-Clang

**Comment by Cyan4973** [comment]
This seems to be more UI behavior than compilation outcome.
Not sure if this issue still needs to be opened.

**Comment by t-mat** [comment]
@boris-999, thank you for the detailed information.

I understand that you've resolved your issue (hence this issue #666).  But you still don't find `<stdalign.h>`.

As for `<stdalign.h>`, only Windows SDK 10.0.20348.0 or later version support it.
So I recommend you to install recent version of Windows SDK via Visual Studio Installer or [their archive](https://developer.microsoft.com/en-us/windows/downloads/sdk-archive/).
I think 10.0.20348.0 would be fine for VS 2019 and toolset v142.

Also you should make sure you're using recent version of Windows SDK.
You can set it at `MSVC 2019 Menu > Project > Property > Configuration Properties > General > (Right Pane) > Windows SDK Version`.

And if this issue #666 is resolved, please close the issue.


> Not sure if this issue still needs to be opened.

@Cyan4973, you're right.  I supposed there's corner case, but it seems there isn't.
Also since GitHub Actions VM for VS2019 has [Windows SDK 10.0.20348.0 and 10.0.22000.0](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md#installed-windows-sdks=), our test doesn't hit `<stdalign.h>` issue.
